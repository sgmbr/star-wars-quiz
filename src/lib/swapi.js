const BASE_URL = 'https://swapi.co/api/'

function checkStatus (response) {
  if (response.status >= 200 && response.status < 300) {
    return Promise.resolve(response)
  } else {
    return Promise.reject(new Error(response.statusText))
  }
}

function json (response) {
  return response.json()
}

function fetchSwapi(url) {
  return (
    fetch(url)
      .then(checkStatus)
      .then(json)
      .catch(err => {
        console.log('Request failed', err)
      })
  )
}

async function getAllEntries (url, entries) {
  entries = entries || []
  const data = await fetchSwapi(url)
  entries = entries.concat(data.results)

  if (data.next) {
    return getAllEntries(data.next, entries)
  } else {
    return entries
  }
}

function getAllEntriesUrls (url) {
  return getAllEntries(url)
          .then(entries => entries.map(entry => entry.url))
}

export { BASE_URL, fetchSwapi, getAllEntriesUrls }
