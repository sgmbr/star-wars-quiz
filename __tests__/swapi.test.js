import { BASE_URL, fetchSwapi, getAllEntriesUrls } from '../src/lib/swapi.js'
import 'whatwg-fetch'

describe('fetchSwapi', () => {
  test('should return Luke Skywalker', () => {
    expect.assertions(1)
    return expect(fetchSwapi(`${BASE_URL}people/1/`)).resolves.toMatchSnapshot()
  })
})

describe('getAllEntriesUrls', () => {
  test('should return all the urls for starships', () => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000
    expect.assertions(1)
    return expect(getAllEntriesUrls(`${BASE_URL}starships`)).resolves.toMatchSnapshot()
  })
})
