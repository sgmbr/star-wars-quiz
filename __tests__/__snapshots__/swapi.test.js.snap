// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchSwapi should return Luke Skywalker 1`] = `
Object {
  "birth_year": "19BBY",
  "created": "2014-12-09T13:50:51.644000Z",
  "edited": "2014-12-20T21:17:56.891000Z",
  "eye_color": "blue",
  "films": Array [
    "https://swapi.co/api/films/2/",
    "https://swapi.co/api/films/6/",
    "https://swapi.co/api/films/3/",
    "https://swapi.co/api/films/1/",
    "https://swapi.co/api/films/7/",
  ],
  "gender": "male",
  "hair_color": "blond",
  "height": "172",
  "homeworld": "https://swapi.co/api/planets/1/",
  "mass": "77",
  "name": "Luke Skywalker",
  "skin_color": "fair",
  "species": Array [
    "https://swapi.co/api/species/1/",
  ],
  "starships": Array [
    "https://swapi.co/api/starships/12/",
    "https://swapi.co/api/starships/22/",
  ],
  "url": "https://swapi.co/api/people/1/",
  "vehicles": Array [
    "https://swapi.co/api/vehicles/14/",
    "https://swapi.co/api/vehicles/30/",
  ],
}
`;

exports[`getAllEntriesUrls should return all the urls 1`] = `
Array [
  "https://swapi.co/api/starships/15/",
  "https://swapi.co/api/starships/5/",
  "https://swapi.co/api/starships/9/",
  "https://swapi.co/api/starships/10/",
  "https://swapi.co/api/starships/11/",
  "https://swapi.co/api/starships/12/",
  "https://swapi.co/api/starships/13/",
  "https://swapi.co/api/starships/21/",
  "https://swapi.co/api/starships/22/",
  "https://swapi.co/api/starships/23/",
  "https://swapi.co/api/starships/27/",
  "https://swapi.co/api/starships/28/",
  "https://swapi.co/api/starships/29/",
  "https://swapi.co/api/starships/31/",
  "https://swapi.co/api/starships/39/",
  "https://swapi.co/api/starships/40/",
  "https://swapi.co/api/starships/41/",
  "https://swapi.co/api/starships/43/",
  "https://swapi.co/api/starships/47/",
  "https://swapi.co/api/starships/48/",
  "https://swapi.co/api/starships/49/",
  "https://swapi.co/api/starships/3/",
  "https://swapi.co/api/starships/59/",
  "https://swapi.co/api/starships/61/",
  "https://swapi.co/api/starships/77/",
  "https://swapi.co/api/starships/17/",
  "https://swapi.co/api/starships/32/",
  "https://swapi.co/api/starships/52/",
  "https://swapi.co/api/starships/58/",
  "https://swapi.co/api/starships/63/",
  "https://swapi.co/api/starships/64/",
  "https://swapi.co/api/starships/65/",
  "https://swapi.co/api/starships/66/",
  "https://swapi.co/api/starships/74/",
  "https://swapi.co/api/starships/75/",
  "https://swapi.co/api/starships/2/",
  "https://swapi.co/api/starships/68/",
]
`;

exports[`getAllEntriesUrls should return all the urls for starships 1`] = `
Array [
  "https://swapi.co/api/starships/15/",
  "https://swapi.co/api/starships/5/",
  "https://swapi.co/api/starships/9/",
  "https://swapi.co/api/starships/10/",
  "https://swapi.co/api/starships/11/",
  "https://swapi.co/api/starships/12/",
  "https://swapi.co/api/starships/13/",
  "https://swapi.co/api/starships/21/",
  "https://swapi.co/api/starships/22/",
  "https://swapi.co/api/starships/23/",
  "https://swapi.co/api/starships/27/",
  "https://swapi.co/api/starships/28/",
  "https://swapi.co/api/starships/29/",
  "https://swapi.co/api/starships/31/",
  "https://swapi.co/api/starships/39/",
  "https://swapi.co/api/starships/40/",
  "https://swapi.co/api/starships/41/",
  "https://swapi.co/api/starships/43/",
  "https://swapi.co/api/starships/47/",
  "https://swapi.co/api/starships/48/",
  "https://swapi.co/api/starships/49/",
  "https://swapi.co/api/starships/3/",
  "https://swapi.co/api/starships/59/",
  "https://swapi.co/api/starships/61/",
  "https://swapi.co/api/starships/77/",
  "https://swapi.co/api/starships/17/",
  "https://swapi.co/api/starships/32/",
  "https://swapi.co/api/starships/52/",
  "https://swapi.co/api/starships/58/",
  "https://swapi.co/api/starships/63/",
  "https://swapi.co/api/starships/64/",
  "https://swapi.co/api/starships/65/",
  "https://swapi.co/api/starships/66/",
  "https://swapi.co/api/starships/74/",
  "https://swapi.co/api/starships/75/",
  "https://swapi.co/api/starships/2/",
  "https://swapi.co/api/starships/68/",
]
`;
